### ISSUE TITLE
[Improvement]: Editor will always fail on printers with the PASSWORD_FEATURE enabled.

### ISSUE THREAD

**Issue by R3velation** [cp2004/OctoPrint-EEPROM-Marlin#50]
### Description of the bug

When connecting to a printer with the password feature enabled, the editor will fail to read from the eeprom and assume that there is no printer connected at all.  This state remains regardless of when the printer is unlocked and only ends when octoprint is disconnects from the printer.  This bug will simply reoccur when the printer re-locks due to octoprint restarting it on connection.

The editor should be able to detect, respond accordingly to a locked printer, and function normally when said printer is put in an unlocked state.  Either by polling to see if the printer is unlocked, or waiting for the user to manually force a retry.

For the sake of possibly saving time (and because its not in any of the data included below), the printer is running marlin 2.0.9.1 and does have a working, enabled eeprom.

### Plugin version

3.1.0

### OctoPrint version

1.6.1

### M503 Response

```gcode
Send: N172 M503*33
Recv: echo:Printer locked! (Unlock with M511 or LCD)
Recv: ok
Send: N173 M511 P1234*87
Recv: ok
[...]
Send: N175 M50*21
Recv: echo:Unknown command: "M50"
Recv: ok
[...]
Send: N177 M503*36
Recv: echo:  G21    ; Units in mm (mm)
Recv: echo:  M149 C ; Units in Celsius
Recv: 
Recv: echo:; Filament settings: Disabled
Recv: echo:  M200 S0 D1.75
Recv: echo:; Steps per unit:
Recv: echo: M92 X80.00 Y80.00 Z80.00 E96.00
Recv: echo:; Maximum feedrates (units/s):
Recv: echo:  M203 X100.00 Y100.00 Z100.00 E100.00
Recv: echo:; Maximum Acceleration (units/s2):
Recv: echo:  M201 X3000.00 Y3000.00 Z3000.00 E3000.00
Recv: echo:; Acceleration (units/s2): P<print_accel> R<retract_accel> T<travel_accel>
Recv: echo:  M204 P3000.00 R3000.00 T3000.00
Recv: echo:; Advanced: B<min_segment_time_us> S<min_feedrate> T<min_travel_feedrate> X<max_x_jerk> Y<max_y_jerk> Z<max_z_jerk> E<max_e_jerk>
Recv: echo:  M205 B20000.00 S0.00 T0.00 X5.00 Y5.00 Z5.00 E5.00
Recv: echo:; Auto Bed Leveling:
Recv: echo:  M420 S0
Recv: echo:; Delta settings: L<diagonal rod> R<radius> H<height> S<segments per sec> XYZ<tower angle trim> ABC<rod trim>
Recv: echo:  M665 L218.00 R97.40 H335.50 S80.00 X0.00 Y0.00 Z0.00 A0.00 B0.00 C0.00
Recv: ; Endstop adjustment:
Recv: echo:  M666 X0.00 Y0.00 Z0.00
Recv: echo:; Material heatup parameters:
Recv: echo:  M145 S0 H200.00 F255
Recv: echo:  M145 S1 H240.00 F255
Recv: echo:; PID settings:
Recv: echo:  M301 P35.54 I3.53 D89.52
Recv: echo:; Filament load/unload lengths:
Recv: echo:  M603 L650.00 U750.00
Recv: ok
```

### API Response

```json
{"backups":[{"name":"eeprom_backup-20210817-172020","time":"2021-08-17 17:20:20"}],"eeprom":{"advanced":{"command":"M205","params":{"B":null,"E":null,"J":null,"S":null,"T":null,"X":null,"Y":null,"Z":null}},"autolevel":{"command":"M420","params":{"S":null,"Z":null}},"bed_pid":{"command":"M304","params":{"D":null,"I":null,"P":null}},"delta":{"command":"M665","params":{"A":null,"B":null,"C":null,"H":null,"L":null,"R":null,"S":null,"X":null,"Y":null,"Z":null}},"endstop":{"command":"M666","params":{"X":null,"Y":null,"Z":null}},"feedrate":{"command":"M203","params":{"E":null,"X":null,"Y":null,"Z":null}},"filament":{"command":"M200","params":{"D":null}},"filament_change":{"command":"M603","params":{"L":null,"U":null}},"filament_runout":{"command":"M412","params":{"D":null,"H":null,"S":null}},"home_offset":{"command":"M206","params":{"X":null,"Y":null,"Z":null}},"hotend_pid":{"command":"M301","params":{"D":null,"I":null,"P":null}},"linear":{"command":"M900","params":{"K":null}},"material1":{"command":"M145","params":{"B":null,"F":null,"H":null,"S":null}},"material2":{"command":"M145","params":{"B":null,"F":null,"H":null,"S":null}},"max_acceleration":{"command":"M201","params":{"E":null,"X":null,"Y":null,"Z":null}},"print_acceleration":{"command":"M204","params":{"P":null,"R":null,"T":null}},"probe_offset":{"command":"M851","params":{"X":null,"Y":null,"Z":null}},"steps":{"command":"M92","params":{"E":null,"X":null,"Y":null,"Z":null}},"tmc_current":{"command":"M906","params":{"E":null,"X":null,"Y":null,"Z":null}},"tmc_hybrid":{"command":"M913","params":{"E":null,"X":null,"Y":null,"Z":null}}},"info":{"additional":{},"capabilities":{},"is_marlin":false,"name":""}}
```

### OctoPrint logs

[octoprint.log](https://github.com/cp2004/OctoPrint-EEPROM-Marlin/files/7004419/octoprint.log)

**Comment by cp2004** [comment]
Once you have unlocked the printer, it should work to hit 'refresh' in the UI to pull in the data.

I won't make the plugin poll the printer, but maybe parsing that message and showing a notice in the UI to unlock your printer first and then refresh would be sensible.

**Comment by cp2004** [comment]
@R3velation - did you get a chance to test the above? Does it work fine after unlocking & refresh? I don't have this feature so I'm not able to test it.

**Comment by R3velation** [comment]
Refreshing after unlocking exhibits the exact same behavior as described in the original post.

**Comment by cp2004** [comment]
When the printer responds that it is locked, a warning will now be shown:

![image](https://user-images.githubusercontent.com/31997505/141662441-84895ffe-d9ba-4344-a460-cd344e2c4e82.png)

When the next M503 command is sent to the printer & goes through OK, the warning is removed - the 'refresh' (& other) buttons are not disabled.

**Comment by cp2004** [comment]
There's a release candidate build available if you want to test it - switch to 'Release Candidate' channel for this plugin in the software update settings in OctoPrint.

Release notes: https://github.com/cp2004/OctoPrint-EEPROM-Marlin/releases/tag/3.2.0rc1
Feedback ticket: #56

**Comment by R3velation** [comment]
This only works when `PASSWORD_ON_STARTUP` is disabled.  The default when `PASSWORD_FEATURE` is enabled is that it will also have `PASSWORD_ON_STARTUP` enabled.

This shows exactly the same behavior as prior.

**Comment by cp2004** [comment]
I'm going to need a full serial.log of the connection then. I wrote the detection based on what you provided in the initial issue, if that's not what always happens then I need the full thing, start to finish. Enable serial.log in the settings, connect to the printer, so whatever you need to do and then upload it here.

**Comment by R3velation** [comment]
[serial.log](https://github.com/cp2004/OctoPrint-EEPROM-Marlin/files/7566828/serial.log)
[octoprint.log](https://github.com/cp2004/OctoPrint-EEPROM-Marlin/files/7566829/octoprint.log)

And in case if you need it, an updated API response.
```json
{"backups":[],"eeprom":{"advanced":{"command":"M205","params":{"B":20000.0,"E":5.0,"J":null,"S":0.0,"T":0.0,"X":5.0,"Y":5.0,"Z":5.0}},"autolevel":{"command":"M420","params":{"S":false,"Z":null}},"bed_pid":{"command":"M304","params":{"D":null,"I":null,"P":null}},"delta":{"command":"M665","params":{"A":0.0,"B":0.0,"C":0.0,"H":335.5,"L":218.0,"R":97.4,"S":80.0,"X":0.0,"Y":0.0,"Z":0.0}},"endstop":{"command":"M666","params":{"X":0.0,"Y":0.0,"Z":0.0}},"feedrate":{"command":"M203","params":{"E":100.0,"X":100.0,"Y":100.0,"Z":100.0}},"filament":{"command":"M200","params":{"D":1.75}},"filament_change":{"command":"M603","params":{"L":650.0,"U":750.0}},"filament_runout":{"command":"M412","params":{"D":null,"H":null,"S":null}},"home_offset":{"command":"M206","params":{"X":null,"Y":null,"Z":null}},"hotend_pid":{"command":"M301","params":{"D":90.22,"I":3.55,"P":35.82}},"linear":{"command":"M900","params":{"K":null}},"material1":{"command":"M145","params":{"B":null,"F":255.0,"H":190.0,"S":0.0}},"material2":{"command":"M145","params":{"B":null,"F":255.0,"H":240.0,"S":1.0}},"max_acceleration":{"command":"M201","params":{"E":3000.0,"X":3000.0,"Y":3000.0,"Z":3000.0}},"print_acceleration":{"command":"M204","params":{"P":3000.0,"R":3000.0,"T":3000.0}},"probe_offset":{"command":"M851","params":{"X":null,"Y":null,"Z":null}},"steps":{"command":"M92","params":{"E":417.0,"X":80.0,"Y":80.0,"Z":80.0}},"tmc_current":{"command":"M906","params":{"E":null,"X":null,"Y":null,"Z":null}},"tmc_hybrid":{"command":"M913","params":{"E":null,"X":null,"Y":null,"Z":null}}},"info":{"additional":{},"capabilities":{},"is_marlin":false,"name":""},"stats":{"failed":0,"filament":"","finished":0,"longest":"","prints":0,"total_time":""}}
```

**Comment by cp2004** [comment]
Ok, I see what the issue is now, thank you.

The printer never responds to OctoPrint's M115 command, which should tell OP all the details about the printer, including capabilities and the firmware name. The EEPROM editor uses the data collected here by OctoPrint to activate the editor, because it can then confirm that the printer is Marlin based.

The plugin also only monitors the serial output when it is collecting data (after an M503/M501 command), to make sure it is not impacting performance at all, but M503 is only sent once we have confirmed the printer is connected & is Marlin, and so it has no knowledge of the printer's locked state. Hopefully you can see the circle created now...

I will have to work out a way of showing that the printer is connected, but locked without depending on knowing it's capabilities to activate the editor first.

**Comment by cp2004** [comment]
I will get to this, eventually - struggling with time. Implemented a printer lock into the virtual printer in OctoPrint so I can fully test it & experiment, hopefully get something good made soon.

**Comment by cp2004** [comment]
Published a new prerelease (3.2.0rc2) - this should work fine now, given it a good test with the actual printer lock feature to test against it is so much easier. Will go to stable as soon as possible - if there's any leftover bugs I can always publish a bugfix release.

**Comment by R3velation** [comment]
Tested RC2, the odd behavior noted in RC1 is no longer present and everything is behaving exactly as it should.
I'd say this issue has been resolved.
